### ISSUE TITLE
BuildOptions generator helper factory

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/build#799]
See https://github.com/dart-lang/json_serializable/blob/bfc6df1af0ec91abc25c9cf3bfc0d763721d843e/json_serializable/lib/builder.dart#L9

You could imagine a generator/builder that, given a function like `jsonPartBuilder` could generate `Builder jsonSerializable(BuilderOptions options) =>` with all of the type checking, etc you'd want

Good errors for wrong types and missing arguments
Warnings about provided, but unsupported arguments, etc.

It'd also be some awesome inception-like magic.

We could even auto-generate the canonical/pretty docs for https://github.com/dart-lang/json_serializable/issues/92

**Comment by jakemac53** [comment]
Agreed that this would be nice - although it can't be done with a builder - or at least not one that is part of the main build.

We could attempt to bake in support in some other form (like using dart:mirrors....).

**Comment by natebosch** [comment]
> although it can't be done with a builder - or at least not one that is part of the main build.

I think the idea would be when you are publishing a Builder you'd run a build step with `build_to: source` and publish it with pub.

**Comment by kevmoo** [comment]
What Nate said

**Comment by jakemac53** [comment]
Ah ya that would work, good point :D

**Comment by davidmorgan** [comment]
Closing as stale.
